Porosity and smoothed porosity changes

GiudGiud opened the pull request:
Pronghorn patch 
https://hpcgitlab.hpc.inl.gov/idaholab/pronghorn/-/merge_requests/864

GiudGiud committed:
Dont replace eps with the smoothed eps, keep both around

refs #21455

lindsayad commented on modules/navier_stokes/src/actions/NSFVAction.C:
```diff
@@ -489,6 +489,7 @@ NSFVAction::NSFVAction(InputParameters parameters)
     _turbulence_handling(getParam<MooseEnum>("turbulence_handling")),
     _porous_medium_treatment(getParam<bool>("porous_medium_treatment")),
     _porosity_name(getParam<MooseFunctorName>("porosity")),
+    _smooth_porosity_name(isParamValid("smoothing_layers") ? NS::smooth_eps : _porosity_name),
```
I would name this data member something different. Maybe `_porosity_functor_name`?

lindsayad commented on modules/navier_stokes/src/userobjects/PINSFVRhieChowInterpolator.C:
```diff
@@ -100,16 +100,6 @@ PINSFVRhieChowInterpolator::pinsfvSetup()
   Moose::FV::interpolateReconstruct(
       _smoothed_eps, _eps, _smoothing_layers, false, _geometric_fi, *this);
   ADReal::do_derivatives = saved_do_derivatives;
-
-  _eps.assign(_smoothed_eps);
-
-  const auto porosity_name = deduceFunctorName(NS::porosity);
-  for (const auto tid : make_range((unsigned int)(1), libMesh::n_threads()))
-  {
-    auto & other_epss = const_cast<Moose::Functor<ADReal> &>(
-        UserObject::_subproblem.getFunctor<ADReal>(porosity_name, tid, name()));
-    other_epss.assign(_smoothed_eps);
-  }
```
If you want to be able to retrieve the smoothed porosity in consumer objects, you will need to add the smoothed porosity to the subproblem's collection of functors.

GiudGiud replied:
yeah and we need to keep track of them in a _epss array too because that s where some constructs retrieve it

should be good now

lindsayad reviewed (changes requested):
If you wanted to, you could now remove the letter-and-envelope system, but understand if you don't want to take that on here

GiudGiud committed:
Address Alex review:
- broadcast functor to all threads
- rename functor name
refs #21455

GiudGiud committed:
Adapt NS not using action to use the smoothed porosity functor, refs #21455

Dont define smoothed porosity functors if there are no smoothing layers
refs #21455

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1111661/) on e704006 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21491/site)

This comment will be updated on new commits.

lindsayad commented:
Something still not quite right when `_smoothiing_layers` is nonzero

lindsayad commented:
If you're not quite sure what's going on here, I can probably resolve it

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1141962/) on 1f80f5f wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>3b8671</th>
      <th colspan=4>#21491 1f80f5</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.41%</td>
      <td>85.37%</td>
      <td>-0.04%</td>
      <td>95.83%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>9725</td>
      <td>9725</td>
      <td>-</td>
      <td>23</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1661</td>
      <td>1667</td>
      <td>+6</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21491/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21491/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21491/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud committed:
Add functor output for debugging
closes #19423

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1178099/) on 445c5ce wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21491/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21491/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a9629f41ab9b99f74b50ed46a5e9c04cbad45f6c`

GiudGiud committed:
Add test for functors debug output. refs #19423

GiudGiud commented:
@lindsayad Pgh patch ready

I added a debug functor functionality. 
It s not the most detailed thing, but it really really helped debugging what was going on with this PR. I was using porosity instead of smoothed_porosity in some objects in Pronghorn & in NSFV action and could not figure it out

lindsayad commented:
what about these other failures?

GiudGiud commented:
Probably unrelated 
I think I have a PR up for that stochastic tool failure we re seeing

lindsayad commented:
Haha did you even look? Two of the failures are from your `show_functors` test

lindsayad commented on modules/navier_stokes/include/userobjects/PINSFVRhieChowInterpolator.h:
```diff
@@ -38,13 +38,19 @@ class PINSFVRhieChowInterpolator : public INSFVRhieChowInterpolator
+  /// interpolations and reconstructions to create the resulting smoothed field
+  const Moose::Functor<ADReal> & _eps;
+
+  /// The smoothed porosity functor/field. After we construct this functor/field we assign it to the
+  /// subproblem's porosity functor
```
```suggestion
  /// The smoothed porosity functor/field
```

lindsayad commented on modules/navier_stokes/src/actions/NSFVAction.C:
```diff
@@ -489,6 +489,8 @@ NSFVAction::NSFVAction(InputParameters parameters)
     _turbulence_handling(getParam<MooseEnum>("turbulence_handling")),
     _porous_medium_treatment(getParam<bool>("porous_medium_treatment")),
     _porosity_name(getParam<MooseFunctorName>("porosity")),
+    _smooth_porosity_functor_name(isParamValid("porosity_smoothing_layers") ? NS::smoothed_porosity
+                                                                            : _porosity_name),
```
I think a name like `_flow_kernel_porosity_functor_name` might be better? Because this may or may not represent a smoothed porosity functor

GiudGiud replied:
_flow_porosity_functor?

lindsayad replied:
i like that

lindsayad commented on modules/navier_stokes/src/userobjects/PINSFVRhieChowInterpolator.C:
```diff
@@ -58,7 +59,19 @@ PINSFVRhieChowInterpolator::PINSFVRhieChowInterpolator(const InputParameters & p
+    {
+      if (!UserObject::_subproblem.hasFunctor(NS::smoothed_porosity, tid))
+        // Smoothed porosity is only an envelope at this point, it will be set during pinsfvSetup()
+        UserObject::_subproblem.addFunctor(NS::smoothed_porosity, _smoothed_eps, tid);
```
At some point maybe we'll want to add a user ability to control what the name is if there's multiple, disconnected fluid flow regions, but even there they may choose to use the same variable sets anyway. So I don't think we need to do anything about this right now

GiudGiud commented on framework/src/problems/SubProblem.C:
```diff
@@ -1039,6 +1047,31 @@ SubProblem::initialSetup()
 }
 
+void
+SubProblem::showFunctors(const THREAD_ID tid) const
```
ok cleaned up

GiudGiud committed:
Apply suggestions from code review
-rename smoothed porosity functor in NSFV action
-remove white space at end of functor lists
Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

GiudGiud commented:
Review should be addressed now

lindsayad commented:
still need to fix those failures

lindsayad reviewed (commented):
My comments have indeed been addressed. Let's just get those failures fixed up and then we can merge!

GiudGiud commented:
ok let s see how this goes

lindsayad reviewed (approved):
beautiful code
